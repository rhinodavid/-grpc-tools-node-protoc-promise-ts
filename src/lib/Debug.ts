import * as LibFs from "fs";
import * as LibUtil from "util";

const LOG_PATH = "/tmp/protoc-promise-gen-ts.debug.log";

export function log(info: any): void {
  LibFs.appendFileSync(
    LOG_PATH,
    LibUtil.inspect(info, { showHidden: true, depth: null }) + "\n"
  );
}
