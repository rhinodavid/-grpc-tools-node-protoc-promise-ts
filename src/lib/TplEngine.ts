import * as fs from "fs";
import * as handlebars from "handlebars";
import * as helpers from "handlebars-helpers";
import * as path from "path";

import { format as prettier } from "prettier";

const TPL_BASE_PATH = path.join(__dirname, "template");

const templateCache = {};

export function registerHelper(
  name: string,
  fn: handlebars.HelperDelegate
): void {
  handlebars.registerHelper(name, fn);
}

export function compile(templateName: string): HandlebarsTemplateDelegate {
  return handlebars.compile(
    fs.readFileSync(`${path.join(TPL_BASE_PATH, templateName)}.hbs`).toString()
  );
}

export function render(
  templateName: string,
  params: { [key: string]: any }
): string {
  const template =
    templateCache[templateName] ||
    (templateCache[templateName] = compile(templateName));
  return template(params);
}

helpers({ handlebars: handlebars });
handlebars.registerHelper("curlyLeft", function () {
  return "{";
});
handlebars.registerHelper("curlyRight", function () {
  return "}";
});
handlebars.registerHelper("render", function (
  templateName: string,
  params: { [key: string]: any }
) {
  const result = render(templateName, params);
  return prettier(result, { parser: "typescript" });
});
