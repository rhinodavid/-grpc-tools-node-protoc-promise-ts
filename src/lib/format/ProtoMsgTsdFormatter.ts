import { EnumModel, format as formatEnum } from "./partial/EnumFormatter";
import {
  ExtensionModel,
  format as formatExtension,
} from "./partial/ExtensionFormatter";
import {
  MessageModel,
  format as formatMessage,
} from "./partial/MessageFormatter";
import {
  filePathFromProtoWithoutExt,
  filePathToPseudoNamespace,
  getPathToRoot,
} from "../Utility";

import { DependencyFilter } from "../DependencyFilter";
import { ExportMap } from "../ExportMap";
import { FileDescriptorProto } from "google-protobuf/google/protobuf/descriptor_pb";
import { WellKnownTypesMap } from "../WellKnown";

export interface ProtoMsgTsdModel {
  packageName: string;
  fileName: string;
  imports: string[];
  messages: MessageModel[];
  extensions: ExtensionModel[];
  enums: EnumModel[];
}

export function format(
  descriptor: FileDescriptorProto,
  exportMap: ExportMap
): ProtoMsgTsdModel {
  const fileName = descriptor.getName();
  const packageName = descriptor.getPackage();

  const imports: Array<string> = [];
  const messages: Array<MessageModel> = [];
  const extensions: Array<ExtensionModel> = [];
  const enums: Array<EnumModel> = [];

  const upToRoot = getPathToRoot(fileName);

  imports.push(`import * as jspb from "google-protobuf";`);
  descriptor.getDependencyList().forEach((dependency: string) => {
    if (DependencyFilter.indexOf(dependency) !== -1) {
      return; // filtered
    }
    const pseudoNamespace = filePathToPseudoNamespace(dependency);
    if (dependency in WellKnownTypesMap) {
      imports.push(
        `import * as ${pseudoNamespace} from "${WellKnownTypesMap[dependency]}";`
      );
    } else {
      const filePath = filePathFromProtoWithoutExt(dependency);
      imports.push(
        `import * as ${pseudoNamespace} from "${upToRoot}${filePath}";`
      );
    }
  });

  descriptor.getMessageTypeList().forEach((enumType) => {
    messages.push(formatMessage(fileName, exportMap, enumType, "", descriptor));
  });
  descriptor.getExtensionList().forEach((extension) => {
    extensions.push(formatExtension(fileName, exportMap, extension, ""));
  });
  descriptor.getEnumTypeList().forEach((enumType) => {
    enums.push(formatEnum(enumType, ""));
  });

  return {
    packageName: packageName,
    fileName: fileName,
    imports: imports,
    messages: messages,
    extensions: extensions,
    enums: enums,
  };
}
