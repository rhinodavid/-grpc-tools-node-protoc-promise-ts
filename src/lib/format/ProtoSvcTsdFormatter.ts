import { MESSAGE_TYPE, getFieldType } from "./partial/FieldTypesFormatter";
import {
  filePathFromProtoWithoutExt,
  filePathToPseudoNamespace,
  getPathToRoot,
} from "../Utility";

import { DependencyFilter } from "../DependencyFilter";
import { ExportMap } from "../ExportMap";
import { FileDescriptorProto } from "google-protobuf/google/protobuf/descriptor_pb";
import { WellKnownTypesMap } from "../WellKnown";
import { registerHelper } from "../TplEngine";

export interface ServiceType {
  serviceName: string;
  methods: Array<ServiceMethodType>;
}

export const defaultServiceType = JSON.stringify({
  serviceName: "",
  methods: [],
} as ServiceType);

export interface ServiceMethodType {
  packageName: string;
  serviceName: string;
  methodName: string;
  requestStream: boolean;
  responseStream: boolean;
  requestTypeName: string;
  responseTypeName: string;
  type: string; // "ClientUnaryCall" || "ClientWritableStream" || "ClientReadableStream" || "ClientDuplexStream"
}

export const defaultServiceMethodType = JSON.stringify({
  packageName: "",
  serviceName: "",
  methodName: "",
  requestStream: false,
  responseStream: false,
  requestTypeName: "",
  responseTypeName: "",
  type: "",
} as ServiceMethodType);

export interface ProtoSvcTsdModel {
  packageName: string;
  fileName: string;
  imports: string[];
  services: Array<ServiceType>;
}

export function format(
  descriptor: FileDescriptorProto,
  exportMap: ExportMap
): ProtoSvcTsdModel {
  if (descriptor.getServiceList().length === 0) {
    return null;
  }

  const fileName = descriptor.getName();
  const packageName = descriptor.getPackage();
  const upToRoot = getPathToRoot(fileName);

  const imports: Array<string> = [];
  const services: Array<ServiceType> = [];

  // Need to import the non-service file that was generated for this .proto file
  imports.push(`import * as grpc from "grpc";`);
  const asPseudoNamespace = filePathToPseudoNamespace(fileName);
  imports.push(
    `import * as ${asPseudoNamespace} from "${upToRoot}${filePathFromProtoWithoutExt(
      fileName
    )}";`
  );

  descriptor.getDependencyList().forEach((dependency: string) => {
    if (DependencyFilter.indexOf(dependency) !== -1) {
      return; // filtered
    }
    const pseudoNamespace = filePathToPseudoNamespace(dependency);
    if (dependency in WellKnownTypesMap) {
      imports.push(
        `import * as ${pseudoNamespace} from "${WellKnownTypesMap[dependency]}";`
      );
    } else {
      const filePath = filePathFromProtoWithoutExt(dependency);
      imports.push(
        `import * as ${pseudoNamespace} from "${upToRoot + filePath}";`
      );
    }
  });

  descriptor.getServiceList().forEach((service) => {
    const serviceData = JSON.parse(defaultServiceType) as ServiceType;

    serviceData.serviceName = service.getName();

    service.getMethodList().forEach((method) => {
      const methodData = JSON.parse(
        defaultServiceMethodType
      ) as ServiceMethodType;

      methodData.packageName = packageName;
      methodData.serviceName = serviceData.serviceName;
      methodData.methodName = method.getName();
      methodData.requestStream = method.getClientStreaming();
      methodData.responseStream = method.getServerStreaming();
      methodData.requestTypeName = getFieldType(
        MESSAGE_TYPE,
        method.getInputType().slice(1),
        "",
        exportMap
      );
      methodData.responseTypeName = getFieldType(
        MESSAGE_TYPE,
        method.getOutputType().slice(1),
        "",
        exportMap
      );

      if (!methodData.requestStream && !methodData.responseStream) {
        methodData.type = "ClientUnaryCall";
      } else if (methodData.requestStream && !methodData.responseStream) {
        methodData.type = "ClientWritableStream";
      } else if (!methodData.requestStream && methodData.responseStream) {
        methodData.type = "ClientReadableStream";
      } else if (methodData.requestStream && methodData.responseStream) {
        methodData.type = "ClientDuplexStream";
      }

      serviceData.methods.push(methodData);
    });

    services.push(serviceData);
  });

  registerHelper("lcFirst", function (str) {
    return str.charAt(0).toLowerCase() + str.slice(1);
  });

  return {
    packageName: packageName,
    fileName: fileName,
    imports: imports,
    services: services,
  };
}
