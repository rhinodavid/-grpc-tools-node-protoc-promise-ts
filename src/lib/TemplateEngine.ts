import * as fs from "fs";
import * as handlebars from "handlebars";
import * as helpers from "handlebars-helpers";
import * as path from "path";

import { ProtoMessageDefinitionModel } from "./format/ProtoMessageDefinitionFormatter";
import { ProtoServiceDefinitionModel } from "./format/ProtoServiceDefinitionFormatter";
import { format as prettier } from "prettier";

const TPL_BASE_PATH = path.join(__dirname, "template");

const templateCache = {};

export function compile(
  templateName: "message_definition_template" | "service_definition_template"
): HandlebarsTemplateDelegate {
  return handlebars.compile(
    fs.readFileSync(`${path.join(TPL_BASE_PATH, templateName)}.hbs`).toString()
  );
}

export function render(
  templateName: "message_definition_template" | "service_definition_template",
  params: ProtoMessageDefinitionModel | ProtoServiceDefinitionModel
): string {
  let template;
  if (templateCache[templateName]) {
    template = templateCache[templateName];
  } else {
    templateCache[templateName] = compile(templateName);
    template = templateCache[templateName];
  }
  const result = template(params);
  return prettier(result, { parser: "typescript" });
}

export function registerHelper(
  name: string,
  fn: handlebars.HelperDelegate
): void {
  handlebars.registerHelper(name, fn);
}

helpers({ handlebars });

handlebars.registerHelper("curlyLeft", function () {
  return "{";
});
handlebars.registerHelper("curlyRight", function () {
  return "}";
});

handlebars.registerHelper("renderPartial", function (
  templateName: "message_definition_template" | "service_definition_template",
  params: ProtoMessageDefinitionModel | ProtoServiceDefinitionModel
) {
  return render(templateName, params);
});
