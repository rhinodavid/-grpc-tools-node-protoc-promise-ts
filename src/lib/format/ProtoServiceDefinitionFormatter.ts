import { MESSAGE_TYPE, getFieldType } from "./partial/FieldTypesFormatter";
import {
  filePathFromProtoWithoutExt,
  filePathToPseudoNamespace,
  getPathToRoot,
} from "../Utility";

import { DependencyFilter } from "../DependencyFilter";
import { ExportMap } from "../ExportMap";
import { FileDescriptorProto } from "google-protobuf/google/protobuf/descriptor_pb";
import { WellKnownTypesMap } from "../WellKnown";
import { registerHelper } from "../TemplateEngine";

export interface ServiceType {
  serviceName: string;
  methods: Array<ServiceMethodType>;
}

export interface ServiceMethodType {
  packageName: string;
  serviceName: string;
  methodName: string;
  requestStream: boolean;
  responseStream: boolean;
  requestTypeName: string;
  responseTypeName: string;
  type: string; // "ClientUnaryCall" || "ClientWritableStream" || "ClientReadableStream" || "ClientDuplexStream"
}

export interface ProtoServiceDefinitionModel {
  dateString: string;
  fileName: string;
  imports: string[];
  generatePromiseClients: boolean;
  packageName: string;
  services: Array<ServiceType>;
}

export function format(
  descriptor: FileDescriptorProto,
  exportMap: ExportMap,
  generatePromiseClients: boolean,
  dateString: string = new Date().toString()
): ProtoServiceDefinitionModel {
  if (descriptor.getServiceList().length === 0) {
    return null;
  }

  const fileName = descriptor.getName();
  const packageName = descriptor.getPackage();
  const upToRoot = getPathToRoot(fileName);

  const imports: Array<string> = [];
  const services: Array<ServiceType> = [];

  // Need to import the non-service file that was generated for this .proto file
  imports.push(`import * as grpc from "grpc";`);
  const asPseudoNamespace = filePathToPseudoNamespace(fileName);
  imports.push(
    `import * as ${asPseudoNamespace} from "${upToRoot}${filePathFromProtoWithoutExt(
      fileName
    )}";`
  );

  descriptor.getDependencyList().forEach((dependency: string) => {
    if (DependencyFilter.indexOf(dependency) !== -1) {
      return; // filtered
    }
    const pseudoNamespace = filePathToPseudoNamespace(dependency);
    if (dependency in WellKnownTypesMap) {
      imports.push(
        `import * as ${pseudoNamespace} from "${WellKnownTypesMap[dependency]}";`
      );
    } else {
      const filePath = filePathFromProtoWithoutExt(dependency);
      imports.push(
        `import * as ${pseudoNamespace} from "${upToRoot + filePath}";`
      );
    }
  });

  descriptor.getServiceList().forEach((service) => {
    const serviceData = {
      serviceName: "",
      methods: [],
    };

    serviceData.serviceName = service.getName();

    service.getMethodList().forEach((method) => {
      const methodData = {
        methodName: "",
        packageName: "",
        requestStream: false,
        requestTypeName: "",
        responseStream: false,
        responseTypeName: "",
        serviceName: "",
        type: "",
      };
      methodData.packageName = packageName;
      methodData.serviceName = serviceData.serviceName;
      methodData.methodName = method.getName();
      methodData.requestStream = method.getClientStreaming();
      methodData.responseStream = method.getServerStreaming();
      methodData.requestTypeName = getFieldType(
        MESSAGE_TYPE,
        method.getInputType().slice(1),
        "",
        exportMap
      );
      methodData.responseTypeName = getFieldType(
        MESSAGE_TYPE,
        method.getOutputType().slice(1),
        "",
        exportMap
      );

      if (!methodData.requestStream && !methodData.responseStream) {
        methodData.type = "ClientUnaryCall";
      } else if (methodData.requestStream && !methodData.responseStream) {
        methodData.type = "ClientWritableStream";
      } else if (!methodData.requestStream && methodData.responseStream) {
        methodData.type = "ClientReadableStream";
      } else if (methodData.requestStream && methodData.responseStream) {
        methodData.type = "ClientDuplexStream";
      }

      serviceData.methods.push(methodData);
    });
    services.push(serviceData);
  });

  registerHelper("lcFirst", function (str) {
    return str.charAt(0).toLowerCase() + str.slice(1);
  });

  return {
    dateString,
    fileName,
    imports,
    generatePromiseClients,
    packageName,
    services,
  };
}
