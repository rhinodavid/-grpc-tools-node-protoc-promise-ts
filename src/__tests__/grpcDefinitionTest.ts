import * as grpcLibrary from "@grpc/grpc-js";
import * as protoLoader from "@grpc/proto-loader";

import {
  ProtobufTypeDefinition,
  ServiceClientConstructor,
} from "@grpc/grpc-js/build/src/make-client";

import { flatten, mapObject } from "underscore";
import { format as formatProtoMessage } from "../lib/format/ProtoMessageDefinitionFormatter";

import { format as formatProtoService } from "../lib/format/ProtoServiceDefinitionFormatter";
import { render } from "../lib/TemplateEngine";
import { ExportMap } from "../lib/ExportMap";
import { FileDescriptorProto } from "google-protobuf/google/protobuf/descriptor_pb";
import { GrpcObject } from "@grpc/grpc-js";

describe("gRCP service definitions", () => {
  describe("generate grpc service definition", () => {
    it("without promise clients", async () => {
      const packageDefinition = await protoLoader.load(
        `${__dirname}/test.proto`,
        {}
      );

      const packageObject: Record<
        string,
        GrpcObject | ServiceClientConstructor | ProtobufTypeDefinition
      > = grpcLibrary.loadPackageDefinition(packageDefinition);

      const isDefs = (
        o: GrpcObject | ServiceClientConstructor | ProtobufTypeDefinition
      ): o is ProtobufTypeDefinition => {
        // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
        return (<ProtobufTypeDefinition>o).fileDescriptorProtos !== undefined;
      };
      const namesToDescriptions = mapObject(
        packageObject,
        (
          x: Record<
            string,
            GrpcObject | ProtobufTypeDefinition | ServiceClientConstructor
          >
        ) =>
          mapObject(x, (o) =>
            isDefs(o)
              ? o.fileDescriptorProtos.map((buffer) => {
                  const typedInputBuffer = new Uint8Array(
                    (buffer as any).length
                  );
                  typedInputBuffer.set(buffer);
                  const proto = FileDescriptorProto.deserializeBinary(
                    typedInputBuffer
                  );
                  return proto;
                })
              : null
          )
      );
      const testDescriptors = namesToDescriptions["test"]; // see what gets rendered for "test.proto"

      const fileDescriptorProtos: Array<FileDescriptorProto> = flatten(
        Object.values(testDescriptors).filter(Boolean)
      );

      // TODO: make sure each descriptor is unique

      const exportMap = new ExportMap();
      fileDescriptorProtos.forEach((b) => exportMap.addFileDescriptor(b));

      const dateString: string = new Date(
        "Tue Sep 21 2020 03:15:20 GMT-0600 (Mountain Daylight Time)"
      ).toString();

      fileDescriptorProtos.forEach((fileDescriptorProto) => {
        const message = render(
          "message_definition_template",
          formatProtoMessage(fileDescriptorProto, exportMap, dateString)
        );
        expect(message).toMatchSnapshot();

        const service = render(
          "service_definition_template",
          formatProtoService(fileDescriptorProto, exportMap, false, dateString)
        );
        expect(service).toMatchSnapshot();

        const serviceWithPromises = render(
          "service_definition_template",
          formatProtoService(fileDescriptorProto, exportMap, true, dateString)
        );
        expect(serviceWithPromises).toMatchSnapshot();
      });
    });
  });
});
