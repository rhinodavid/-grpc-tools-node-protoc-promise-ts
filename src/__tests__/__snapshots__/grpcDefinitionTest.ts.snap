// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gRCP service definitions generates grpc service definition: test.proto_message 1`] = `
"/**
 *       ***************************************
 *          AUTOGENERATED FILE - DO NOT EDIT
 *       ***************************************
 *
 * This file was generated by grpc-tools-node-protoc-promise-ts
 * https://github.com/rhinodavid/grpc-tools-node-protoc-promise-ts
 *
 * package: test
 * file: test.proto
 * generated at: Mon Sep 21 2020 03:15:20 GMT-0600 (Mountain Daylight Time)
 */

/* tslint:disable */
/* eslint-disable */

import * as jspb from \\"google-protobuf\\";
export class MessageA extends jspb.Message {
  getInt(): number;
  setInt(value: number): MessageA;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): MessageA.AsObject;
  static toObject(includeInstance: boolean, msg: MessageA): MessageA.AsObject;
  static extensions: { [key: number]: jspb.ExtensionFieldInfo<jspb.Message> };
  static extensionsBinary: {
    [key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>;
  };
  static serializeBinaryToWriter(
    message: MessageA,
    writer: jspb.BinaryWriter
  ): void;
  static deserializeBinary(bytes: Uint8Array): MessageA;
  static deserializeBinaryFromReader(
    message: MessageA,
    reader: jspb.BinaryReader
  ): MessageA;
}

export namespace MessageA {
  export type AsObject = {
    pb_int: number;
  };
}

export class MessageB extends jspb.Message {
  getInt(): number;
  setInt(value: number): MessageB;

  serializeBinary(): Uint8Array;
  toObject(includeInstance?: boolean): MessageB.AsObject;
  static toObject(includeInstance: boolean, msg: MessageB): MessageB.AsObject;
  static extensions: { [key: number]: jspb.ExtensionFieldInfo<jspb.Message> };
  static extensionsBinary: {
    [key: number]: jspb.ExtensionFieldBinaryInfo<jspb.Message>;
  };
  static serializeBinaryToWriter(
    message: MessageB,
    writer: jspb.BinaryWriter
  ): void;
  static deserializeBinary(bytes: Uint8Array): MessageB;
  static deserializeBinaryFromReader(
    message: MessageB,
    reader: jspb.BinaryReader
  ): MessageB;
}

export namespace MessageB {
  export type AsObject = {
    pb_int: number;
  };
}
"
`;

exports[`gRCP service definitions generates grpc service definition: test.proto_promise_service 1`] = `
"/**
 *       ***************************************
 *          AUTOGENERATED FILE - DO NOT EDIT
 *       ***************************************
 *
 * This file was generated by grpc-tools-node-protoc-promise-ts
 * https://github.com/rhinodavid/grpc-tools-node-protoc-promise-ts
 *
 * package: test
 * file: test.proto
 * generated at: Mon Sep 21 2020 03:15:20 GMT-0600 (Mountain Daylight Time)
 */

/* tslint:disable */
/* eslint-disable */

import * as grpc from \\"grpc\\";
import * as test_pb from \\"./test_pb\\";
import { TDuplexRpc, TUnaryRpc } from \\"grpc-promise-ts\\";

interface ITestService
  extends grpc.ServiceDefinition<grpc.UntypedServiceImplementation> {
  duplex: ITestService_Iduplex;
  duplex: ITestService_Iduplex;
  duplex: ITestService_Iduplex;
  duplex: ITestService_Iduplex;
}

interface ITestService_Iduplex
  extends grpc.MethodDefinition<test_pb.MessageA, test_pb.MessageB> {
  path: string; // \\"/test.Test/duplex\\"
  requestStream: boolean; // true
  responseStream: boolean; // true
  requestSerialize: grpc.serialize<test_pb.MessageA>;
  requestDeserialize: grpc.deserialize<test_pb.MessageA>;
  responseSerialize: grpc.serialize<test_pb.MessageB>;
  responseDeserialize: grpc.deserialize<test_pb.MessageB>;
}
interface ITestService_Iduplex
  extends grpc.MethodDefinition<test_pb.MessageA, test_pb.MessageB> {
  path: string; // \\"/test.Test/duplex\\"
  requestStream: boolean; // true
  responseStream: boolean; // true
  requestSerialize: grpc.serialize<test_pb.MessageA>;
  requestDeserialize: grpc.deserialize<test_pb.MessageA>;
  responseSerialize: grpc.serialize<test_pb.MessageB>;
  responseDeserialize: grpc.deserialize<test_pb.MessageB>;
}
interface ITestService_Iduplex
  extends grpc.MethodDefinition<test_pb.MessageA, test_pb.MessageB> {
  path: string; // \\"/test.Test/duplex\\"
  requestStream: boolean; // true
  responseStream: boolean; // true
  requestSerialize: grpc.serialize<test_pb.MessageA>;
  requestDeserialize: grpc.deserialize<test_pb.MessageA>;
  responseSerialize: grpc.serialize<test_pb.MessageB>;
  responseDeserialize: grpc.deserialize<test_pb.MessageB>;
}
interface ITestService_Iduplex
  extends grpc.MethodDefinition<test_pb.MessageA, test_pb.MessageB> {
  path: string; // \\"/test.Test/duplex\\"
  requestStream: boolean; // true
  responseStream: boolean; // true
  requestSerialize: grpc.serialize<test_pb.MessageA>;
  requestDeserialize: grpc.deserialize<test_pb.MessageA>;
  responseSerialize: grpc.serialize<test_pb.MessageB>;
  responseDeserialize: grpc.deserialize<test_pb.MessageB>;
}

export const TestService: ITestService;

export interface ITestServer {
  duplex: grpc.handleBidiStreamingCall<test_pb.MessageA, test_pb.MessageB>;
  duplex: grpc.handleBidiStreamingCall<test_pb.MessageA, test_pb.MessageB>;
  duplex: grpc.handleBidiStreamingCall<test_pb.MessageA, test_pb.MessageB>;
  duplex: grpc.handleBidiStreamingCall<test_pb.MessageA, test_pb.MessageB>;
}

export interface ITestClient {
  duplex(): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    options: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    metadata: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    options: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    metadata: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    options: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    metadata: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    options: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    metadata: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
}

export class TestClient extends grpc.Client implements ITestClient {
  constructor(
    address: string,
    credentials: grpc.ChannelCredentials,
    options?: object
  );
  public duplex(
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    metadata?: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    metadata?: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    metadata?: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    metadata?: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
}

/**
 * Interface and class to declare a Test client that has been promisifyed with the
 * \`grpc-promise-ts\` (https://github.com/rhinodavid/grpc-promise-ts) package
 */
export interface ITestPromiseClient {
  duplex: TDuplexRpc<test_pb.MessageA, test_pb.MessageB>;
  duplex: TDuplexRpc<test_pb.MessageA, test_pb.MessageB>;
  duplex: TDuplexRpc<test_pb.MessageA, test_pb.MessageB>;
  duplex: TDuplexRpc<test_pb.MessageA, test_pb.MessageB>;
}

export class TestPromiseClient extends grpc.Client
  implements ITestPromiseClient {
  // This class cannot be directly instantiated.
  // To get an instance of TestPromiseClient, first create
  // a TestClient and then pass it to the \`convertToPromiseClient\`
  // function in the \`gprc-promise-ts\` package
  public duplex: TDuplexRpc<test_pb.MessageA, test_pb.MessageB>;
  public duplex: TDuplexRpc<test_pb.MessageA, test_pb.MessageB>;
  public duplex: TDuplexRpc<test_pb.MessageA, test_pb.MessageB>;
  public duplex: TDuplexRpc<test_pb.MessageA, test_pb.MessageB>;

  // TODO: add other types of RPCs
}
"
`;

exports[`gRCP service definitions generates grpc service definition: test.proto_service 1`] = `
"/**
 *       ***************************************
 *          AUTOGENERATED FILE - DO NOT EDIT
 *       ***************************************
 *
 * This file was generated by grpc-tools-node-protoc-promise-ts
 * https://github.com/rhinodavid/grpc-tools-node-protoc-promise-ts
 *
 * package: test
 * file: test.proto
 * generated at: Mon Sep 21 2020 03:15:20 GMT-0600 (Mountain Daylight Time)
 */

/* tslint:disable */
/* eslint-disable */

import * as grpc from \\"grpc\\";
import * as test_pb from \\"./test_pb\\";

interface ITestService
  extends grpc.ServiceDefinition<grpc.UntypedServiceImplementation> {
  duplex: ITestService_Iduplex;
  duplex: ITestService_Iduplex;
}

interface ITestService_Iduplex
  extends grpc.MethodDefinition<test_pb.MessageA, test_pb.MessageB> {
  path: string; // \\"/test.Test/duplex\\"
  requestStream: boolean; // true
  responseStream: boolean; // true
  requestSerialize: grpc.serialize<test_pb.MessageA>;
  requestDeserialize: grpc.deserialize<test_pb.MessageA>;
  responseSerialize: grpc.serialize<test_pb.MessageB>;
  responseDeserialize: grpc.deserialize<test_pb.MessageB>;
}
interface ITestService_Iduplex
  extends grpc.MethodDefinition<test_pb.MessageA, test_pb.MessageB> {
  path: string; // \\"/test.Test/duplex\\"
  requestStream: boolean; // true
  responseStream: boolean; // true
  requestSerialize: grpc.serialize<test_pb.MessageA>;
  requestDeserialize: grpc.deserialize<test_pb.MessageA>;
  responseSerialize: grpc.serialize<test_pb.MessageB>;
  responseDeserialize: grpc.deserialize<test_pb.MessageB>;
}

export const TestService: ITestService;

export interface ITestServer {
  duplex: grpc.handleBidiStreamingCall<test_pb.MessageA, test_pb.MessageB>;
  duplex: grpc.handleBidiStreamingCall<test_pb.MessageA, test_pb.MessageB>;
}

export interface ITestClient {
  duplex(): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    options: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    metadata: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    options: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  duplex(
    metadata: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
}

export class TestClient extends grpc.Client implements ITestClient {
  constructor(
    address: string,
    credentials: grpc.ChannelCredentials,
    options?: object
  );
  public duplex(
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    metadata?: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
  public duplex(
    metadata?: grpc.Metadata,
    options?: Partial<grpc.CallOptions>
  ): grpc.ClientDuplexStream<test_pb.MessageA, test_pb.MessageB>;
}
"
`;
